### ISSUE TITLE
Azure AD login dependency on Azure DevOps service (conditional access block)

### ISSUE THREAD

**Issue by glmacedo** [microsoft/AzureStorageExplorer#7701]
### Preflight Checklist

- [X] I have installed the [latest version of Storage Explorer](https://github.com/Microsoft/AzureStorageExplorer/releases/latest).
- [X] I have checked existing resources, including the [troubleshooting guide](https://docs.microsoft.com/azure/storage/common/storage-explorer-troubleshooting) and the [release notes](https://github.com/microsoft/AzureStorageExplorer/releases).
- [X] I have searched for [similar issues](https://github.com/Microsoft/AzureStorageExplorer/issues).

### Storage Explorer Version

1.32.1

### Architecture

x64

### Storage Explorer Build Number

20231114.10

### Platform

Windows

### OS Version

Windows 10+

### Bug Description

Azure Storage Explorer Azure AD (EntraID) authentication is getting denied due to conditional policy for "Azure DevOps" application.

Conditional access policy states that Azure DevOps requires 3 factor authentication (password, token, managed device). Azure Storage Explorer login is failing from unmanaged device due to this policy.

Authentication succeeds from the same device with both Azure CLI and AZCopy.

### Steps to Reproduce

1. Create conditional access policy requiring 3 factors (password, token, managed device) for Azure DevOps application.
2. Launch Storage Explorer and try to authenticate with Azure AD (EntraID).
3. Browser window will open prompting for authentication (account, password) and token/authenticator flow.
4. After MFA succeeds, access will be denied due to conditional access policy created in step 1.

### Actual Experience

OAUTH authentication denied due to conditional policy applied to Azure DevOps application - which should have no impact to Azure Storage Explorer itself.

### Expected Experience

Authentication successful with two factors (password and token/authenticator).

### Additional Context

Authentication succeeds when access is done from a managed device that is capable of satisfying all 3 factors (password, token/authenticator, device state).

The failure from "unmanaged" devices creates an issue for OT devices that need to store data on blob but are not fully managed (due to OT limitations).

**Comment by MRayermannMSFT** [comment]
@glmacedo thanks for reaching out. We've been made aware of your request both here and from the folks in customer support you are talking to. Since you opened this issue, we're happy to interact with you here. Can you please let us know/share:

1. Are you failing to sign in with a work/school account (WSA) or a microsoft account (MSA)? It is ok if you don't know.
2. Can you please follow these steps (https://learn.microsoft.com/en-us/troubleshoot/azure/azure-storage/storage-explorer-troubleshooting?tabs=Windows#authentication-logs) to gather **Authentication Logs**. Please make sure to follow all the steps, including enabling **Verbose Authentication Logging**, and ideally clearing the contents of the ServiceHub/logs folder before starting the Storage Explorer session where you reproduce the sign in failure.

Thanks!

**Comment by glmacedo** [comment]
@MRayermannMSFT, thanks for the response. Please find answers below:

1. Yes, failure happens with "work/school" account.
2. Authentication logs are attached to this message.

As further context, working with the MSFT Technical Advisor for Azure Storage, we identified that commenting the lines for "VisualStudioOnlineAudience" and "VisualStudioOnlineEndpoint" under the  _defaultAadProviderSettings/azure section in "AzureEnvironmentsManagement.js" I was able to get AAD WSA authentication working even with the conditional access policy for Azure DevOps.

![image](https://github.com/microsoft/AzureStorageExplorer/assets/44760079/dd36d664-ebc3-496b-9b6e-200d8611717a)


The Technical Advisor explained that this is an unsupported configuration that may be removed in the future and that this setting is in place to account for MSA - would it be possible to implement this "choice" as a feature for use with WSA?

Thanks,

G.
[AzureStorageExplorer-authenticationlogs.zip](https://github.com/microsoft/AzureStorageExplorer/files/14086459/AzureStorageExplorer-authenticationlogs.zip)


``

**Comment by MRayermannMSFT** [comment]
@glmacedo can you confirm if when you try to sign in:

1. Your web browser opens
2. A page opens where you have to enter your username/password, maybe peform multifactor authentication
3. **Important**: After succesfully doing the basic steps of authenticating (aka step 2), you see a page like this:

![image](https://github.com/microsoft/AzureStorageExplorer/assets/21112697/36e7af0b-f8aa-4fa6-bea0-da18d5f6d174)

Does that all match your experience?

**Comment by glmacedo** [comment]
No, the steps are as follows:
1 - a browser window opens asking for my user account.
2 - I enter the WSA account and the password
3 - A prompt for my Authenticator tokens open up
4 - I authorize it on my authenticator app
5 - Authenticator app succeeds and the web page briefly flashes to "you're authenticated" but then quickly changes to a page saying I can't have access to the resource (due to the Azure DevOps dependency, as my machine doesn't match the 3 factors that are required, as it is an unmanaged device). The resulting page is shown on the screenshot below
![image](https://github.com/microsoft/AzureStorageExplorer/assets/44760079/cc442531-a615-4084-83eb-ce9c2c0042c8)

If I comment out the lines I mentioned above on my response, authentication happens without a problem and I can then  sign on to the resources that are made available with AAD authentication (either with public or private endpoints, depending on where I'm connecting from).

Thanks,

G.

**Comment by MRayermannMSFT** [comment]
Thanks for the answers @gimacedo ! We've prepared a private build which contains changes that should unblock this for you. Can you please download the build, install, and try to sign in? When you try to sign in, please make sure to click "OK" on the page you shared. As long as you do that, the sign in should work (even if you see a message saying otherwise).

If it doesn't work, please do go ahead and collect those authentication logs again.

Build link: https://storageexplorerpublish.blob.core.windows.net/privatebuilds/7701/StorageExplorer-windows-x64.exe?sv=2023-01-03&st=2024-01-29T17%3A52%3A00Z&se=2024-02-29T17%3A52%3A00Z&sr=b&sp=r&sig=b7qJN9s0QjO%2FqU1I4wUNJJo%2FCnjK4xnMHNk9Te%2FPbjA%3D

**Comment by MRayermannMSFT** [comment]
@glmacedo great! Glad to hear it worked. Yes, we will include the change in our next release (1.33). That is still a few weeks away though, for now, please go back to using 1.32.1 (https://github.com/microsoft/AzureStorageExplorer/releases/tag/v1.32.1), with the workaround you discussed earlier (https://github.com/microsoft/AzureStorageExplorer/issues/7701#issuecomment-1914974123). We do not recommend using private builds beyond their original purpose (in this case verifying that the sign in fix worked).

**Comment by MRayermannMSFT** [comment]
Closing since the fix is merged into main. As previously stated, it will be released as part of 1.33. Until then, users of 1.32 can do the workaround as helpfully described here: https://github.com/microsoft/AzureStorageExplorer/issues/7701#issuecomment-1914974123
